src/nifreeze/estimator.py

```diff
@@ -22,229 +22,151 @@
 #
 """A model-based algorithm for the realignment of dMRI data."""
 
+from __future__ import annotations
+
 from pathlib import Path
-from tempfile import TemporaryDirectory, mkstemp
+from tempfile import TemporaryDirectory
+from typing import Self
 
-import nibabel as nb
 from tqdm import tqdm
 
-from nifreeze.data.splitting import lovo_split
-from nifreeze.model.base import ModelFactory
-from nifreeze.registration.ants import _prepare_registration_data, _run_registration
+from nifreeze.data.base import BaseDataset
+from nifreeze.model.base import BaseModel, ModelFactory
+from nifreeze.registration.ants import (
+    _prepare_registration_data,
+    _run_registration,
+)
 from nifreeze.utils import iterators
 
 
+class Filter:
+    """Alters an input data object (e.g., downsampling)."""
+
+    def run(self, dataset: BaseDataset, **kwargs):
+        """
+        Trigger execution of the designated filter.
+
+        Parameters
+        ----------
+        dataset : :obj:`~nifreeze.data.base.BaseDataset`
+            The input dataset this estimator operates on.
+
+        Returns
+        -------
+        :obj:`~nifreeze.estimator.Estimator`
+            The estimator, after fitting.
+
+        """
+        return dataset
```

Comment:
```suggestion
    def run(self, dataset: BaseDataset, **kwargs) -> BaseDataset:
        """
        Trigger execution of the designated filter.

        Parameters
        ----------
        dataset : :obj:`~nifreeze.data.base.BaseDataset`
            The input dataset this estimator operates on.

        Returns
        -------
        dataset : :obj:`~nifreeze.data.base.BaseDataset`
            The dataset, after filtering.

        """
        return dataset
```

Really, you want to do something like:

```python
DatasetT = ty.TypeVar('DatasetT', bound=BaseDataset)

class Filter:
    def run(self, dataset: DatasetT, **kwargs) -> DatasetT:
        return dataset
```

This ensures that, if `pet_ds` is a `PETDataset`, then a `Filter().run(pet_ds)` will be understood to return a `PETDataset`, and likewise any other dataset type.